Blog: Units of Measurement blog post

htreu opened the pull request:
This accompanies the new QuantityType PR #4818 with a high level blog post showing the relevant parts for unit based state updates and commands.

Signed-off-by: Henning Treu <henning.treu@telekom.de>

triller-telekom commented on docs/_posts/2018-02-22-units-of-measurement.md:
```diff
@@ -0,0 +1,109 @@
+A lot of functionality in a smart home setup is based on environmental conditions which until now are measured without an explicit unit. Now Eclipse SmartHome comes with full "Units of Measurement" support to make the unit part of the item state and provide automatic or specific unit conversion. <!--more--> 
+
+##### Overview
+A lot of functionality in a smart home setup is based on environmental conditions. Sensors will report temperatures from all sorts of locations, other weather data is collected and displayed or in house luminance is measured to control steady lighting conditions.
```
~~other~~ weather data...

displayed **,** or

triller-telekom commented on docs/_posts/2018-02-22-units-of-measurement.md:
```diff
@@ -0,0 +1,109 @@
+
+##### Overview
+A lot of functionality in a smart home setup is based on environmental conditions. Sensors will report temperatures from all sorts of locations, other weather data is collected and displayed or in house luminance is measured to control steady lighting conditions.
+All these sensor data is not only measured with a scalar value which represents the current state but has an implicit unit attached to it: Temperatures are usually given in either centigrade or degree Fahrenheit, wind speed will be displayed in either kilometers per hour (km/h) or miles per hour (mph) and luminance is given in Lux or it may even be a device specific unit or value scale.
```
Shorter sentences would improve the readability here.

triller-telekom commented on docs/_posts/2018-02-22-units-of-measurement.md:
```diff
@@ -0,0 +1,109 @@
+##### Overview
+A lot of functionality in a smart home setup is based on environmental conditions. Sensors will report temperatures from all sorts of locations, other weather data is collected and displayed or in house luminance is measured to control steady lighting conditions.
+All these sensor data is not only measured with a scalar value which represents the current state but has an implicit unit attached to it: Temperatures are usually given in either centigrade or degree Fahrenheit, wind speed will be displayed in either kilometers per hour (km/h) or miles per hour (mph) and luminance is given in Lux or it may even be a device specific unit or value scale.
+Up to now these units where not represented in Eclipse SmartHome and the solution/user had to "know" its units and model items and rules accordingly.
```
"solution" is kind of irritating here, I would remove it.

triller-telekom commented on docs/_posts/2018-02-22-units-of-measurement.md:
```diff
@@ -0,0 +1,109 @@
+A lot of functionality in a smart home setup is based on environmental conditions. Sensors will report temperatures from all sorts of locations, other weather data is collected and displayed or in house luminance is measured to control steady lighting conditions.
+All these sensor data is not only measured with a scalar value which represents the current state but has an implicit unit attached to it: Temperatures are usually given in either centigrade or degree Fahrenheit, wind speed will be displayed in either kilometers per hour (km/h) or miles per hour (mph) and luminance is given in Lux or it may even be a device specific unit or value scale.
+Up to now these units where not represented in Eclipse SmartHome and the solution/user had to "know" its units and model items and rules accordingly.
+In order to unify the different values and units and let the smart home make use of it, Eclipse SmartHome now comes with support for "Units of Measurement" or in short "uom".
```
short "uom". -> short "UoM".

Though, I am not 100% sure about that. But I think its a defined term.

triller-telekom commented on docs/_posts/2018-02-22-units-of-measurement.md:
```diff
@@ -0,0 +1,109 @@
+Depending on the users locale or cultural preference a specific unit might fit the users needs. For temperature it is common to use centigrade (°C) in most places around the globe but degree Fahrenheit (°F) is used in the United States of America.
+
+##### UoM data with QuantityType
+Sensor devices and internet services connected to Eclipse SmartHome as Things all provide their scalar values in an implicit unit of measurement. The binding developer knows what exact unit the sensor or service data uses and can now pass those values to the framework using the new [`QuantityType`](../documentation/concepts/units-of-measurement.html). QuantityType represents the scalar value with ...
```
as Things ~~all~~ provide

and can ~~now~~ pass those

triller-telekom commented on docs/_posts/2018-02-22-units-of-measurement.md:
```diff
@@ -0,0 +1,109 @@
+
+##### UoM data with QuantityType
+Sensor devices and internet services connected to Eclipse SmartHome as Things all provide their scalar values in an implicit unit of measurement. The binding developer knows what exact unit the sensor or service data uses and can now pass those values to the framework using the new [`QuantityType`](../documentation/concepts/units-of-measurement.html). QuantityType represents the scalar value with ...
+The Yahoo Weather binding makes use of QuantityType and publishes state updates in the following form:
```
`QuantityType` <- hightlight it

triller-telekom commented on docs/_posts/2018-02-22-units-of-measurement.md:
````diff
@@ -0,0 +1,109 @@
+```
+
+In the temperature example the state is marked to be a "<Temperature>" typed QuantitType. It will only accept units which correspond to the dimension type, like centigrade, degree Fahrenheit or Kelvin.
+In the same way the Thing`s Channels and the linked Number items have to be typed to match against each other. Both Channel and Item definition use an extended item type format which passes the expected dimension to a Number item.
````
In the same way the Thing's <- wrong ' this will break the formatting.

triller-telekom commented on docs/_posts/2018-02-22-units-of-measurement.md:
```diff
@@ -0,0 +1,109 @@
+In the temperature example the state is marked to be a "<Temperature>" typed QuantitType. It will only accept units which correspond to the dimension type, like centigrade, degree Fahrenheit or Kelvin.
+In the same way the Thing`s Channels and the linked Number items have to be typed to match against each other. Both Channel and Item definition use an extended item type format which passes the expected dimension to a Number item.
+
+The temperature Channel definition for Yahoo Weather looks like this:
```
The temperature channel...

triller-telekom commented on docs/_posts/2018-02-22-units-of-measurement.md:
````diff
@@ -0,0 +1,109 @@
+</channel-type>
+```
+
+A corresponding item definition in the *.items DSL is then:
````
I would highlight `*.items`

A corresponding item definition in the `*.items` DSL is ~~then~~:

triller-telekom commented on docs/_posts/2018-02-22-units-of-measurement.md:
````diff
@@ -0,0 +1,109 @@
+```
+
+Both definitions use the extended item type `Number:Temperature` which configures the Number item to only accept state updates with units from the dimension "Temperature".
+For backwards compatibility a Number item without a dimension specified will receive state updates as DecimalType. For this situation the scalar value from the QuantityType is taken without any conversion and passed on to the item. 
````
For backward~~s~~ compatibility

...without a specified dimension will...

triller-telekom commented on docs/_posts/2018-02-22-units-of-measurement.md:
```diff
@@ -0,0 +1,109 @@
+Once a binding provides state updates with QuantityTypes the unit which should be used on the item can be different from the unit selected by the binding.
+
+###### Locale Conversion
+The default behaviour of the framework is to convert the source value to the unit which is defined to be the default for the current measurement system.
```
behaviour: are we using british or american english? I thought we are using the later one, so: behavior

triller-telekom commented on docs/_posts/2018-02-22-units-of-measurement.md:
```diff
@@ -0,0 +1,109 @@
+Apart from the locale based selection the framework may be configured to use a predefined measurement system by configuring the `UnitProvider` component.
+
+###### Item State Conversion
+In case the default conversion to the system's measurement system does not fit the current use case items may define their own target unit. The formatter pattern in the item's state description may use the `%unit%` placeholder to render the default unit (see item definition example above) or just define a fixed unit to which state updates will be converted:
```
fit the current use case **,** items may

triller-telekom commented on docs/_posts/2018-02-22-units-of-measurement.md:
```diff
@@ -0,0 +1,109 @@
+With this definition the target unit is defined to be Kelvin (K) so all state updates to this item will be converted to Kelvin.
+
+##### Rendering Units in UIs
+The same unit conversion mechanism used through the item's state description can be used in sitemap descriptions. The `label` property of a widget in a sitemap and also the `mapping` property of Switch and Select widgets support the unit definition in the formatter pattern. 
```
used by the item's

`Switch` and `Select` <- highlight them

triller-telekom commented on docs/_posts/2018-02-22-units-of-measurement.md:
```diff
@@ -0,0 +1,109 @@
+
+##### Scripts & Rules
+The classic rule engine in Eclipse SmartHome also has full support of the new QuantityType. It can be used in arithmetic calculations and comparison with automatic unit conversion.
+The `Weather_Temperature` item from the example above and fixed unit Kelvin can be compared against other QuantityType instances with other units from the Temperature dimension.
```
example above **with the** fixed unit

htreu commented:
Thanks @triller-telekom for the review. All comments considered valid and fixed :-)
I also changed "Kelvin" to "kelvin" from insights of [this discussion](https://english.stackexchange.com/questions/173906/should-units-of-measure-be-capitalized).

sjka commented on docs/_posts/2018-02-22-units-of-measurement.md:
```diff
@@ -0,0 +1,109 @@
+title: "Units of Measurement with Eclipse SmartHome"
+date:   2018-02-22
+image: "2018-02-22-units-of-measurement.png"
+published: false
```
`true`please, `true`!

htreu replied:
Thanks, done :-)

J-N-K commented:
You might think of adding the package names that we need to import. Eclipse is not able to chose the ones for MILLI or HECTO automatically.

htreu commented:
@J-N-K thanks for the comment, I added a short paragraph with the respective package names. This should then also be added to the official documentation.
@sjka speaking of official docs: We need a release of the latest version.

sjka commented:
Will do it anyway when publishing this

sjka commented:
> This should then also be added to the official documentation.

@htreu Would you please do so?

htreu commented:
@sjka done.

triller-telekom commented on docs/documentation/concepts/units-of-measurement.md:
```diff
@@ -65,3 +65,9 @@ In order to match `NumberItems` and channels and define a default state descript
+When creating QuantityType states the framework offers some useful packages and classes:
+The `org.eclipse.smarthome.core.library.unit` package contains the classes `SIUnits`, `ImperialUnits` and `SmartHomeUnits` which provide units unique to either of the measurement systems and common units used in both systems.
+The `MetricPrefix` class provides prefixes like MILLI, CENTI, HECTO, etc. which are wrappers to create derived units.
+The `org.eclipse.smarthome.core.library.dimension` and `javax.measure.quantity` packages provide interface which are used to type the generic QuantityType and units. 
```
interface**s**

maggu2810 commented on docs/_posts/2018-02-22-units-of-measurement.md:
````diff
@@ -0,0 +1,118 @@
+updateState(pressureChannelUID, state);
+```
+
+In the temperature example the state is marked to be a "<Temperature>" typed QuantitType. It will only accept units which correspond to the dimension type, like degree Celsius, degree Fahrenheit or kelvin.
````
Does `"<Temperature>"` work for you?
At least Github's Markdown parser cannot display it correctly.

htreu replied:
good find, thanks @maggu2810.
